flax/linen/module.py

```diff
@@ -519,6 +519,22 @@ def reimport(self, other: '_ModuleInternalState') -> None:
 capture_call_intermediates = lambda _, method_name: method_name == '__call__'
 
 
+class ParentDescriptor:
+  """Wraps parent module references in weak refs.
+
+  This prevents reference cycles from forming via parent links which can lead
+  to accidental OOMs in eager mode due to slow garbage collection as well as
+  spurious tracer leaks during jit compilation.
+  """
+  def __get__(self, obj, objtype=None):
+    parent = object.__getattribute__(obj, "_parent_ref")
```

Comment:
Isn't this the same as:

```
@property
def parent(self):
  parent = self.__getattribute__(obj, "_parent_ref")
  return parent() if isinstance(parent, weakref.ReferenceType) else parent
 
@parent.setter
def parent(self, value):
   maybe_weak = weakref.ref(value) if isinstance(value, Module) else value
    object.__setattr__(self, "_parent_ref", maybe_weak)
```

This seems simpler than summoning the dark power of Python descriptors?

Reply:
Ha, it would be, but at least for now I have to keep `cls.parent` set to a `dataclasses.field` for the dataclass transform, and _then_ replace the `cls.parent` with a descriptor during `__init_subclass__`.   There's a separate PR that's going to refactor our custom dataclass transform machinery, and I might be able to use `property` instead once that's in.

Reply:
Makes sense, let's go with the descriptor for now then